docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+    Compare(a, b []byte) (int, error)
+    // CompareString returns an integer comparing the two strings. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    CompareString(a, b string) (int, error)
+    // Key returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    Key(buf *Buffer, str []byte) error
+    // KeyFromString returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    KeyFromString(buf *Buffer, str string) error
+
+    /******* Possible Optimization ********/
+    // FromKey returns the original byte slice from collation key.
+    FromKey(buf *Buffer, str []byte) error
+    // StringFromKey returns the original string from collation key.
+    StringFromKey(str []byte) string
+}
+```
+The interface is quite similar to the Go [collate package](https://godoc.org/golang.org/x/text/collate), it is possible to implement some of the MySQL collation by simply mapping it to an existing Go collation.
+
+### Row Format
+
+The encoding layout of TiDB has been described in our [previous article](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map). The row format should be changed to make it memory comparable, this is important to the index lookup. Basic principle is that all keys encoded should use the `sortKeys` result from `Key()`/`KeyFromString()` function. However, most of the `sortKeys` calculations are not reversible.
+
+#### Option 1
+
+  * For table data, encodings stay unchanged. All strings are compared after decoding with the `Compare()` function.
+  * For table indices, we replace current `ColumnValue` with `sortKey`:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID
+	```
+    - For non-unique indices:
+    ```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: null
+	```
+
+Pros:
+    The current encoding layout keeps unchanged.
+
+Cons:
+     Since index value can not be recovered from `sortKey`, extra table lookup is needed for current Index-only queries, this may lead to performance regressions. The extra table lookup also requires corresponding changes to the TiDB optimizer.
+
+#### Option 2
+
+  * Table data encodings is the same with Option 1.
+  * For table indices, we encode the `ColumnValue` to the value of the storage, so as to avoid extra table lookup for index queries:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID_indexedColumnsValue
+	```
+    - For non-unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: indexedColumnsValue
+	```
+
+Pros: Extra table lookup is avoided and no change to the TIDB optimizer.
+
+Cons: The size of index value on string column with new collations is doubled(we need to write both `sortKey` and `indexedColumnsValue`, their sizes are equal for most of the collations).
+
```

Comment:
Option 3 is to store the rowID in collation order rather than byte order, but not store the sortKey. I assume that this is not possible, but it might be worth putting it here and saying "not possible" -- so we can compare to how it works in MySQL.

Reply:
> I assume that this is not possible.

This is not possible, the rowID of a record in TiDB is a unique BIGINT that is allocated randomly(the only exception is the case when the primary key of a table is a single integer column, value of this column is used as rowID). 

And the rowID must be unique, right? It is not possible by the definition.

Reply:
What's the benefit if it is possible?